cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -221,75 +290,76 @@ private void processMetrics() {
+              }
             } catch (IOException e) {
               LOG.warn("Failed to decode message to MetricValue. Skipped. {}", e.getMessage());
             }
           }
         }
 
+        if (currentMessageId != null) {
+          topicMessageIds.put(topicIdMetaKey, currentMessageId);
+          lastMessageId = currentMessageId;
+        }
+
+        // Skip persisting if the records is empty
         if (records.isEmpty()) {
-          LOG.trace("No metrics record to process.");
           return;
         }
 
-        persistRecords(records);
-        records.clear();
-        lastMessageId = nextMessageId;
-        persistMessageId(lastMessageId);
-      } catch (Exception e) {
-        LOG.error("Failed to process metrics. Will be retried in next iteration.", e);
-      }
-    }
-
-    private void persistRecords(List<MetricValues> metricValues) throws Exception {
-      long now = System.currentTimeMillis();
-      addProcessingStats(metricValues, now);
-      metricStore.add(metricValues);
-      recordsProcessed += metricValues.size();
-      // avoid logging more than once a minute
-      if (now > lastLoggedMillis + TimeUnit.MINUTES.toMillis(1)) {
-        lastLoggedMillis = now;
-        LOG.debug("{} metrics records processed in thread {}. Last record time: {}.",
-                 recordsProcessed, getName(), metricValues.get(metricValues.size() - 1).getTimestamp());
-      }
-    }
-
-    private void addProcessingStats(List<MetricValues> records, long currentTime) {
-      int count = records.size();
-      long delay = currentTime - TimeUnit.SECONDS.toMillis(records.get(records.size() - 1).getTimestamp());
-      records.add(
-        new MetricValues(metricsContextMap, TimeUnit.MILLISECONDS.toSeconds(currentTime),
-                         ImmutableList.of(new MetricValue("metrics.process.count", MetricType.COUNTER, count),
-                                          new MetricValue("metrics.process.delay.ms", MetricType.GAUGE, delay))));
-    }
-
-    private void persistMessageId(byte[] messageId) {
-      try {
-        metaTable.save(topicIdMetaKey, messageId);
-        LOG.debug("Persisted last processed MessageId: {} in thread {}", Bytes.toStringBinary(messageId), getName());
+        // Ensure there's only one thread can persist records and messageId's when the persistingFlag is false.
+        // The thread set persistingFlag to true when it starts to persist.
+        if (!persistingFlag.compareAndSet(false, true)) {
+          LOG.trace("Cannot persist because persistingFlag is taken by another thread.");
+          return;
+        }
+        
+        try {
+          LinkedList<MetricValues> recordsCopy = new LinkedList<>();
+          Map<TopicIdMetaKey, byte[]> topicMessageIdsCopy = new HashMap<>(topicMessageIds);
+          // TODO: (CDAP-8327) there is a risk of running out-of-memory if other threads keep writing to records
+          while (!records.isEmpty()) {
```

Comment:
Better not to use `isEmpty` check in a loop for concurrent queue. Use a iterator is better:

```java
Iterator<MetricValues> iterator = records.iterator();
while (iterator.hasNext()) {
  recordsCopy.add(iterator.next());
  iterator.remove();
}
```